File: docs/examples/javascript/vdp_collocation.js
//
//     MIT No Attribution
//
//     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
//
//     Permission is hereby granted, free of charge, to any person obtaining a copy of this
//     software and associated documentation files (the "Software"), to deal in the Software
//     without restriction, including without limitation the rights to use, copy, modify,
//     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
//     permit persons to whom the Software is furnished to do so.
//
//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//
// JS port of docs/examples/python/vdp_collocation.py.
//
// Van der Pol OCP transcribed by hand with Radau collocation; the NLP
// variable vector V is sliced manually (no integrator plugin) -> portable.
//
// JS notes (see README.md):
//   * numpy poly1d/polyder/polyint are replaced by small array helpers.
//   * matplotlib output is dropped; we log the optimum instead.
//   * The Python `expand=True` and `ipopt.linear_solver='ma27'` options
//     are dropped (defaults used; expand crashes the wasm GC teardown).

// Polynomial helpers (coeffs highest-degree-first, like numpy.poly1d).
function polymul(a, b) {
  const r = new Array(a.length + b.length - 1).fill(0);
  for (let i = 0; i < a.length; i++)
    for (let j = 0; j < b.length; j++) r[i + j] += a[i] * b[j];
  return r;
}
function polyval(p, x) { let r = 0; for (const c of p) r = r * x + c; return r; }
function polyder(p) {
  const n = p.length - 1, r = [];
  for (let i = 0; i < n; i++) r.push(p[i] * (n - i));
  return r.length ? r : [0];
}
function polyint(p) {
  const n = p.length, r = [];
  for (let i = 0; i < n; i++) r.push(p[i] / (n - i));
  r.push(0);
  return r;
}

async function example(ca, log) {
  const inf = Infinity;

  // Degree of interpolating polynomial
  const d = 3;

  // Choose collocation points
  const tau_root = [0, ...ca.collocation_points(d, "radau")];

  // Coefficients of the collocation (C), continuity (D), quadrature (F) eqs
  const C = Array.from({ length: d + 1 }, () => new Array(d + 1).fill(0));
  const D = new Array(d + 1).fill(0);
  const Fc = new Array(d + 1).fill(0);

  // Construct polynomial basis
  for (let j = 0; j <= d; j++) {
    let p = [1];
    for (let r = 0; r <= d; r++)
      if (r !== j) p = polymul(p, [1, -tau_root[r]]).map((c) => c / (tau_root[j] - tau_root[r]));
    D[j] = polyval(p, 1.0);
    const pder = polyder(p);
    for (let r = 0; r <= d; r++) C[j][r] = polyval(pder, tau_root[r]);
    Fc[j] = polyval(polyint(p), 1.0);
  }

  // Control discretization
  const nk = 20;
  const tf = 10.0;     // End time
  const h = tf / nk;   // Size of the finite elements

  // Declare variables (use scalar graph)
  const t = ca.SX.sym("t");
  const u = ca.SX.sym("u");
  const x = ca.SX.sym("x", 2);
  const [x0e, x1e] = ca.vertsplit(x);

  // ODE rhs function and quadratures
  const xdot = ca.vertcat(
    ca.plus(ca.minus(ca.times(ca.minus(1, ca.times(x1e, x1e)), x0e), x1e), u),
    x0e);
  const qdot = ca.plus(ca.plus(ca.times(x0e, x0e), ca.times(x1e, x1e)), ca.times(u, u));
  const f = ca.Function("f", [t, x, u], [xdot, qdot], ["t", "x", "u"], ["xdot", "qdot"]);

  // Control bounds
  const u_min = -0.75, u_max = 1.0, u_init = 0.0;

  // State bounds and initial guess
  const x_min = [-inf, -inf], x_max = [inf, inf];
  const xi_min = [0.0, 1.0], xi_max = [0.0, 1.0];
  const xf_min = [0.0, 0.0], xf_max = [0.0, 0.0];
  const x_init = [0.0, 0.0];

  const nx = 2, nu = 1;

  // Total number of variables
  const NX = nk * (d + 1) * nx;  // Collocated states
  const NU = nk * nu;            // Parametrized controls
  const NXF = nx;                // Final state
  const NV = NX + NU + NXF;

  // NLP variable vector
  const V = ca.MX.sym("V", NV);
  const Vs = ca.vertsplit(V); // scalar entries

  // All variables with bounds and initial guess
  const vars_lb = new Array(NV).fill(0);
  const vars_ub = new Array(NV).fill(0);
  const vars_init = new Array(NV).fill(0);
  let offset = 0;

  // Get collocated states and parametrized control as MX sub-vectors
  const X = Array.from({ length: nk + 1 }, () => new Array(d + 1));
  const U = new Array(nk);
  const slice = (from, n) => ca.vcat(Vs.slice(from, from + n));
  for (let k = 0; k < nk; k++) {
    for (let j = 0; j <= d; j++) {
      X[k][j] = slice(offset, nx);
      vars_init.splice(offset, nx, ...x_init);
      if (k === 0 && j === 0) {
        vars_lb.splice(offset, nx, ...xi_min); vars_ub.splice(offset, nx, ...xi_max);
      } else {
        vars_lb.splice(offset, nx, ...x_min); vars_ub.splice(offset, nx, ...x_max);
      }
      offset += nx;
    }
    U[k] = slice(offset, nu);
    vars_lb[offset] = u_min; vars_ub[offset] = u_max; vars_init[offset] = u_init;
    offset += nu;
  }
  // State at end time
  X[nk][0] = slice(offset, nx);
  vars_lb.splice(offset, nx, ...xf_min); vars_ub.splice(offset, nx, ...xf_max);
  vars_init.splice(offset, nx, ...x_init);
  offset += nx;

  // Constraints and objective
  const g = [], lbg = [], ubg = [];
  let J = ca.MX(0);

  for (let k = 0; k < nk; k++) {
    for (let j = 1; j <= d; j++) {
      // State derivative at the collocation point
      let xp_jk = ca.times(C[0][j], X[k][0]);
      for (let r = 1; r <= d; r++) xp_jk = ca.plus(xp_jk, ca.times(C[r][j], X[k][r]));

      // Collocation equations
      const Tkj = h * (k + tau_root[j]);
      const [ode_jk, ltermv] = f(ca.DM(Tkj), X[k][j], U[k]);
      g.push(ca.minus(ca.times(h, ode_jk), xp_jk));
      lbg.push(0, 0); ubg.push(0, 0);

      // Objective contribution
      J = ca.plus(J, ca.times(Fc[j], ca.times(ltermv, h)));
    }
    // State at the end of the finite element
    let xf_k = ca.times(D[0], X[k][0]);
    for (let r = 1; r <= d; r++) xf_k = ca.plus(xf_k, ca.times(D[r], X[k][r]));

    // Continuity equation
    g.push(ca.minus(X[k + 1][0], xf_k));
    lbg.push(0, 0); ubg.push(0, 0);
  }

  const G = ca.vcat(g);
  const nlp = { x: V, f: J, g: G };
  const solver = ca.nlpsol("solver", "ipopt", nlp);

  const res = solver.call({
    x0: ca.DM(vars_init), lbx: ca.DM(vars_lb), ubx: ca.DM(vars_ub),
    lbg: ca.DM(lbg), ubg: ca.DM(ubg),
  });

  log("-----");
  log("optimal cost: " + res["f"]);

  // Get values at the beginning of each finite element (Python: v_opt[0::stride])
  const v_opt = res["x"];
  const stride = (d + 1) * nx + nu;
  log("x0 trajectory = " + v_opt["0::" + stride]);
  log("x1 trajectory = " + v_opt["1::" + stride]);
  log("u  trajectory = " + v_opt[(d + 1) * nx + "::" + stride]);
}

if (typeof require !== "undefined" && typeof module !== "undefined" && require.main === module) {
  const path = require("path");
  const casadiPath = process.env.CASADI_JS
    || path.resolve(__dirname, "../../../build-wasm/swig/wasm-js/casadi.js");
  require(casadiPath)()
    .then((ca) => example(ca, (...a) => console.log(...a)))
    .catch((e) => { console.error("FATAL:", e.message || e); process.exit(1); });
}

if (typeof module !== "undefined" && module.exports) module.exports = example;

File: docs/examples/julia/vdp_collocation.jl
#
#     MIT No Attribution
#
#     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
#
#     Permission is hereby granted, free of charge, to any person obtaining a copy of this
#     software and associated documentation files (the "Software"), to deal in the Software
#     without restriction, including without limitation the rights to use, copy, modify,
#     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#     permit persons to whom the Software is furnished to do so.
#
#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# Van der Pol OCP by direct collocation on Radau points, assembling the
# collocation NLP by hand from one MX decision vector V. numpy poly1d helpers
# become small array-based polynomial routines (coeffs highest-degree-first).
const _jl_dir = get(ENV, "CASADI_JL",
    joinpath(@__DIR__, "..", "..", "..", "julia-proto", "casadi-contact"))
include(joinpath(_jl_dir, "CasADiNative.jl"))
import .CasADiNative as ca

function polymul(a, b)
    r = zeros(length(a) + length(b) - 1)
    for i in 1:length(a), j in 1:length(b)
        r[i + j - 1] += a[i] * b[j]
    end
    r
end
polyval(p, x) = foldl((r, c) -> r * x + c, p; init = 0.0)
polyder(p) = (n = length(p) - 1; n == 0 ? [0.0] : [p[i] * (n - i + 1) for i in 1:n])
polyint(p) = vcat([p[i] / (length(p) - i + 1) for i in 1:length(p)], 0.0)

function main()
    # Degree of interpolating polynomial
    d = 3

    # Choose collocation points (Radau)
    tau_root = [0.0; ca.collocation_points(d, "radau")]

    # Coefficients of the collocation, continuity and quadrature equations
    Cc = zeros(d + 1, d + 1)
    D = zeros(d + 1)
    F = zeros(d + 1)

    # Construct polynomial basis
    for j in 1:(d + 1)
        p = [1.0]
        for r in 1:(d + 1)
            if r != j
                p = polymul(p, [1.0, -tau_root[r]]) ./ (tau_root[j] - tau_root[r])
            end
        end
        D[j] = polyval(p, 1.0)
        pder = polyder(p)
        for r in 1:(d + 1)
            Cc[j, r] = polyval(pder, tau_root[r])
        end
        F[j] = polyval(polyint(p), 1.0)
    end

    # Control discretization
    nk = 20
    tf = 10.0          # End time
    h = tf / nk        # Size of the finite elements

    # Declare variables (use scalar graph)
    t = ca.SX.sym("t")   # time
    u = ca.SX.sym("u")   # control
    x = ca.SX.sym("x", 2)  # state

    # ODE rhs function and quadratures
    xdot = ca.vertcat((1 - x[2] * x[2]) * x[1] - x[2] + u, x[1])
    qdot = x[1] * x[1] + x[2] * x[2] + u * u
    f = ca.Function("f", [t, x, u], [xdot, qdot])

    # Control bounds
    u_min, u_max, u_init = -0.75, 1.0, 0.0

    # State bounds and initial guess
    x_min = [-Inf, -Inf]; x_max = [Inf, Inf]
    xi_min = [0.0, 1.0];  xi_max = [0.0, 1.0]
    xf_min = [0.0, 0.0];  xf_max = [0.0, 0.0]
    x_init = [0.0, 0.0]

    # Dimensions
    nx, nu = 2, 1

    # Total number of variables
    NX = nk * (d + 1) * nx   # Collocated states
    NU = nk * nu             # Parametrized controls
    NXF = nx                 # Final state
    NV = NX + NU + NXF

    # NLP variable vector and bounds/guess
    V = ca.MX.sym("V", NV)
    vars_lb = zeros(NV); vars_ub = zeros(NV); vars_init = zeros(NV)
    offset = 0

    # Get collocated states (as 1-based [k][j]) and parametrized control [k]
    Xc = [Vector{Any}(undef, d + 1) for _ in 1:(nk + 1)]
    U = Vector{Any}(undef, nk)
    for k in 1:nk
        # Collocated states
        for j in 1:(d + 1)
            Xc[k][j] = V[(offset + 1):(offset + nx)]
            vars_init[(offset + 1):(offset + nx)] = x_init
            if k == 1 && j == 1
                vars_lb[(offset + 1):(offset + nx)] = xi_min
                vars_ub[(offset + 1):(offset + nx)] = xi_max
            else
                vars_lb[(offset + 1):(offset + nx)] = x_min
                vars_ub[(offset + 1):(offset + nx)] = x_max
            end
            offset += nx
        end
        # Parametrized controls
        U[k] = V[(offset + 1):(offset + nu)]
        vars_lb[offset + 1] = u_min; vars_ub[offset + 1] = u_max
        vars_init[offset + 1] = u_init
        offset += nu
    end
    # State at end time
    Xc[nk + 1][1] = V[(offset + 1):(offset + nx)]
    vars_lb[(offset + 1):(offset + nx)] = xf_min
    vars_ub[(offset + 1):(offset + nx)] = xf_max
    vars_init[(offset + 1):(offset + nx)] = x_init

    # Constraint function for the NLP
    g = ca.MX[]; J = ca.MX(ca.DM(0.0))

    # For all finite elements
    for k in 1:nk
        # For all collocation points
        for j in 2:(d + 1)
            # State derivative at the collocation point
            xp_jk = ca.MX(ca.DM(zeros(nx)))
            for r in 1:(d + 1)
                xp_jk = xp_jk + Cc[r, j] * Xc[k][r]
            end
            # Collocation equations
            Tkj = h * (k - 1 + tau_root[j])
            fk, qk = f(Tkj, Xc[k][j], U[k])
            push!(g, h * fk - xp_jk)
            # Contribution to objective
            J = J + F[j] * qk * h
        end
        # State at the end of the finite element
        xf_k = ca.MX(ca.DM(zeros(nx)))
        for r in 1:(d + 1)
            xf_k = xf_k + D[r] * Xc[k][r]
        end
        # Continuity equation
        push!(g, Xc[k + 1][1] - xf_k)
    end

    # NLP
    nlp = Dict("x" => V, "f" => J, "g" => ca.vertcat(g))

    # Allocate an NLP solver (mumps default; the python ma27 needs HSL)
    solver = ca.nlpsol("solver", "ipopt", nlp, Dict("expand" => true))

    # Solve (constraints are all equalities -> lbg = ubg = 0)
    res = solver(x0 = vars_init, lbx = vars_lb, ubx = vars_ub, lbg = 0, ubg = 0)

    println("optimal cost: ", Float64(res["f"]))

    # States at the start of each finite element (stride (d+1)*nx + nu)
    stride = (d + 1) * nx + nu
    x0_opt = res["x"][1:stride:end]
    x1_opt = res["x"][2:stride:end]
    println("x[0] at element starts = ", x0_opt)
    println("x[1] at element starts = ", x1_opt)
end

main()
